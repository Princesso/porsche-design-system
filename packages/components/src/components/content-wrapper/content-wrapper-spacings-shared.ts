import type { ContentWrapperWidth } from './content-wrapper-utils';
import { gridGap, gridSafeZoneBase, gridSafeZoneXXL, gridWidthMax } from '@porsche-design-system/utilities-v2';

const oneColumnWidthS = `calc((100% - ${gridSafeZoneBase} * 2 - ${gridGap} * 13) / 14)`;
const oneColumnWidthXXL = `calc((min(100%, ${gridWidthMax}) - ${gridSafeZoneXXL} * 2 - ${gridGap} * 13) / 14)`;
export const offsetHorizontalXXL = `max(0px, (100% - ${gridWidthMax}) / 2)`;

export type GridSpacingWidth = Exclude<ContentWrapperWidth, 'full' | 'fluid'>;
export type GridSpacingDataForWidth = [string, Partial<{ [key in 's' | 'xxl']: string }>];
export type GridSpacingMap = { [key in GridSpacingWidth]: GridSpacingDataForWidth };

export const gridSpacingMapConfig: GridSpacingMap = {
  narrow: [
    gridSafeZoneBase,
    {
      s: `calc(${gridSafeZoneBase} + ${gridGap} * 3 + ${oneColumnWidthS} * 3)`,
      xxl: `calc(${offsetHorizontalXXL} + ${gridSafeZoneXXL} + ${gridGap} * 3 + ${oneColumnWidthXXL} * 3)`,
    },
  ],
  basic: [
    gridSafeZoneBase,
    {
      s: `calc(${gridSafeZoneBase} + ${gridGap} + ${oneColumnWidthS})`,
      xxl: `calc(${offsetHorizontalXXL} + ${gridSafeZoneXXL} + ${gridGap} + ${oneColumnWidthXXL})`,
    },
  ],
  extended: [
    gridSafeZoneBase,
    {
      xxl: `calc(${offsetHorizontalXXL} + ${gridSafeZoneXXL})`,
    },
  ],
};

export const getSpacingForWidth = (gridSpacingMapKey: GridSpacingWidth): GridSpacingDataForWidth =>
  gridSpacingMapConfig[gridSpacingMapKey];
