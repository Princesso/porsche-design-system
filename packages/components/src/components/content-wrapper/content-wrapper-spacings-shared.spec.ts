import { getSpacingForWidth } from './content-wrapper-spacings-shared';

describe('getSpacingForWidth()', () => {
  it.each<Parameters<typeof getSpacingForWidth>>([['narrow'], ['basic'], ['extended']])(
    'should return correct spacing for width: %s',
    (...args) => {
      expect(getSpacingForWidth(...args)).toMatchSnapshot();
    }
  );
});
