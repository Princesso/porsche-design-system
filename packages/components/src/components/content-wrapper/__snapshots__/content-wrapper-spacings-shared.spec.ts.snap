// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSpacingForWidth() should return correct spacing for width: basic 1`] = `
[
  "clamp(22px, 10.625vw - 12px, 192px)",
  {
    "s": "calc(clamp(22px, 10.625vw - 12px, 192px) + clamp(16px, 3.6vw, 36px) + calc((100% - clamp(22px, 10.625vw - 12px, 192px) * 2 - clamp(16px, 3.6vw, 36px) * 13) / 14))",
    "xxl": "calc(max(0px, (100% - 2560px) / 2) + clamp(192px, 50vw - 768px, 512px) + clamp(16px, 3.6vw, 36px) + calc((min(100%, 2560px) - clamp(192px, 50vw - 768px, 512px) * 2 - clamp(16px, 3.6vw, 36px) * 13) / 14))",
  },
]
`;

exports[`getSpacingForWidth() should return correct spacing for width: extended 1`] = `
[
  "clamp(22px, 10.625vw - 12px, 192px)",
  {
    "xxl": "calc(max(0px, (100% - 2560px) / 2) + clamp(192px, 50vw - 768px, 512px))",
  },
]
`;

exports[`getSpacingForWidth() should return correct spacing for width: narrow 1`] = `
[
  "clamp(22px, 10.625vw - 12px, 192px)",
  {
    "s": "calc(clamp(22px, 10.625vw - 12px, 192px) + clamp(16px, 3.6vw, 36px) * 3 + calc((100% - clamp(22px, 10.625vw - 12px, 192px) * 2 - clamp(16px, 3.6vw, 36px) * 13) / 14) * 3)",
    "xxl": "calc(max(0px, (100% - 2560px) / 2) + clamp(192px, 50vw - 768px, 512px) + clamp(16px, 3.6vw, 36px) * 3 + calc((min(100%, 2560px) - clamp(192px, 50vw - 768px, 512px) * 2 - clamp(16px, 3.6vw, 36px) * 13) / 14) * 3)",
  },
]
`;
